File: backend/app/models/financing.py
import uuid
from datetime import date

from sqlalchemy import Date, Enum, ForeignKey, Integer, Numeric, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, UUIDMixin
from app.db.types import GUID
from app.models.enums import (
    AmortizationType,
    CommitmentInstallmentStatus,
    CommitmentStatus,
    CommitmentType,
)


class FinancialCommitment(UUIDMixin, TimestampMixin, Base):
    """Generaliza financiamento / empréstimo / consórcio (Financial Commitment)."""

    __tablename__ = "financial_commitments"

    user_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    institution_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("financial_institutions.id", ondelete="RESTRICT"), nullable=False
    )
    type: Mapped[CommitmentType] = mapped_column(
        Enum(CommitmentType, native_enum=False, length=16), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    asset_value: Mapped[str | None] = mapped_column(Numeric(15, 2), nullable=True)
    down_payment: Mapped[str | None] = mapped_column(Numeric(15, 2), nullable=True)
    financed_amount: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    interest_rate: Mapped[str | None] = mapped_column(Numeric(9, 6), nullable=True)
    installments_total: Mapped[int] = mapped_column(Integer, nullable=False)
    default_installment_amount: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    due_day: Mapped[int] = mapped_column(Integer, nullable=False)
    outstanding_balance: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    status: Mapped[CommitmentStatus] = mapped_column(
        Enum(CommitmentStatus, native_enum=False, length=16),
        default=CommitmentStatus.ATIVO,
        nullable=False,
    )
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    installments: Mapped[list["CommitmentInstallment"]] = relationship(back_populates="commitment")


class CommitmentInstallment(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "commitment_installments"
    __table_args__ = (
        UniqueConstraint("commitment_id", "number", name="uq_commitment_installment_number"),
    )

    commitment_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("financial_commitments.id", ondelete="CASCADE"), nullable=False, index=True
    )
    number: Mapped[int] = mapped_column(Integer, nullable=False)
    due_date: Mapped[date] = mapped_column(Date, nullable=False)
    original_amount: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    updated_amount: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    paid_amount: Mapped[str | None] = mapped_column(Numeric(15, 2), nullable=True)
    payment_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    status: Mapped[CommitmentInstallmentStatus] = mapped_column(
        Enum(CommitmentInstallmentStatus, native_enum=False, length=16),
        default=CommitmentInstallmentStatus.PENDENTE,
        nullable=False,
    )
    transaction_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID, ForeignKey("transactions.id", ondelete="SET NULL"), nullable=True
    )
    amortization_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID, ForeignKey("amortizations.id", ondelete="SET NULL"), nullable=True
    )

    commitment: Mapped["FinancialCommitment"] = relationship(back_populates="installments")


class Amortization(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "amortizations"

    user_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    commitment_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("financial_commitments.id", ondelete="CASCADE"), nullable=False, index=True
    )
    date: Mapped[date] = mapped_column(Date, nullable=False)
    paid_amount: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    nominal_amortized_amount: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    discount_obtained: Mapped[str] = mapped_column(Numeric(15, 2), nullable=False)
    type: Mapped[AmortizationType] = mapped_column(
        Enum(AmortizationType, native_enum=False, length=32), nullable=False
    )
    account_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("accounts.id", ondelete="RESTRICT"), nullable=False
    )
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    transaction_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID, ForeignKey("transactions.id", ondelete="SET NULL"), nullable=True
    )

    affected_installments: Mapped[list["AmortizationInstallment"]] = relationship(
        back_populates="amortization", cascade="all, delete-orphan"
    )


class AmortizationInstallment(UUIDMixin, Base):
    """Associação N:N explícita: rastreia exatamente quais parcelas cada amortização afetou."""

    __tablename__ = "amortization_installments"
    __table_args__ = (
        UniqueConstraint(
            "amortization_id", "commitment_installment_id", name="uq_amortization_installment"
        ),
    )

    amortization_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("amortizations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    commitment_installment_id: Mapped[uuid.UUID] = mapped_column(
        GUID, ForeignKey("commitment_installments.id", ondelete="CASCADE"), nullable=False, index=True
    )

    amortization: Mapped["Amortization"] = relationship(back_populates="affected_installments")
    commitment_installment: Mapped["CommitmentInstallment"] = relationship()

File: frontend/src/api/financing.ts
import { api } from "@/lib/api"
import type {
  Amortization,
  AmortizationType,
  CommitmentInstallment,
  CommitmentType,
  FinancialCommitment,
} from "@/types"

export interface CommitmentInput {
  institution_id: string
  type: CommitmentType
  name: string
  asset_value?: number | null
  down_payment?: number | null
  financed_amount: number
  interest_rate?: number | null
  installments_total: number
  default_installment_amount: number
  start_date: string
  due_day: number
  note?: string | null
}

export interface AmortizationInput {
  date: string
  paid_amount: number
  type: AmortizationType
  account_id: string
  installment_numbers?: number[] | null
  note?: string | null
}

export const financingApi = {
  list: () => api.get<FinancialCommitment[]>("/financing").then((r) => r.data),
  get: (id: string) => api.get<FinancialCommitment>(`/financing/${id}`).then((r) => r.data),
  create: (payload: CommitmentInput) => api.post<FinancialCommitment>("/financing", payload).then((r) => r.data),
  listInstallments: (id: string) =>
    api.get<CommitmentInstallment[]>(`/financing/${id}/installments`).then((r) => r.data),
  payInstallment: (
    commitmentId: string,
    installmentId: string,
    payload: { payment_date: string; account_id: string; paid_amount?: number | null },
  ) =>
    api
      .post<CommitmentInstallment>(
        `/financing/${commitmentId}/installments/${installmentId}/pay`,
        payload,
      )
      .then((r) => r.data),
  undoInstallmentPayment: (commitmentId: string, installmentId: string) =>
    api
      .post<CommitmentInstallment>(`/financing/${commitmentId}/installments/${installmentId}/undo-payment`)
      .then((r) => r.data),
}

export const amortizationsApi = {
  list: (commitmentId: string) =>
    api.get<Amortization[]>(`/amortizations/commitments/${commitmentId}`).then((r) => r.data),
  create: (commitmentId: string, payload: AmortizationInput) =>
    api.post<Amortization>(`/amortizations/commitments/${commitmentId}`, payload).then((r) => r.data),
}
